File: apple-helper/Sources/apple-helper/Protocol.swift
//
//  Protocol.swift
//  apple-helper
//
//  Swift `Codable` mirror of the Apple Helper IPC wire protocol.
//
//  SINGLE SOURCE OF TRUTH lives on the Rust side, alongside the shared fixtures:
//  `frontend/src-tauri/src/apple/fixtures/*.json`. These types MUST decode those
//  exact fixture files byte-for-byte.
//
//  Unlike the ari-notch sidecar (snake_case), this protocol is camelCase, and
//  each response carries a DISTINCT `type` discriminator (`probeResult`,
//  `error`) — there is no single `"response"` tag.
//
//  Field names are spelled explicitly as CodingKeys (rather than relying on a
//  key strategy) so the mapping is auditable against the fixtures.
//
//  Forward-compatibility: any unrecognized `type` decodes to `.unknown` and
//  never throws.
//
//  Both enums provide decode AND encode so the tests can round-trip
//  symmetrically. In practice the sidecar DECODES requests and ENCODES
//  responses.
//

import Foundation

// MARK: - Inbound (Rust core → sidecar)

/// Requests sent from the Rust core down to the sidecar. Phase 1 subset.
enum AppleRequest: Equatable {
    /// Availability check — reply with `.probeResult`.
    case probe
    /// On-device summarization via FoundationModels — reply with `.summarizeResult`
    /// on success or `.error` on any failure.
    case summarize(text: String, instruction: String, maxTokens: Int)
    /// On-device speech-to-text of ONE complete audio segment via
    /// SpeechAnalyzer/SpeechTranscriber — reply with `.transcribeResult` on
    /// success or `.error` on any failure. `pcmBase64` is base64 of raw
    /// little-endian Float32, 16 kHz, mono PCM.
    case transcribe(pcmBase64: String, locale: String)
    /// On-device text embeddings via NaturalLanguage `NLEmbedding` — reply with
    /// `.embedResult` (one vector per input, in order) on success or `.error` on
    /// any failure.
    case embedBatch(texts: [String])
    /// Clean shutdown — the sidecar should `exit(0)`.
    case shutdown
    /// Install on-device model assets and STREAM real download progress. Unlike
    /// every other request, this yields MULTIPLE response lines: zero or more
    /// `.progress` lines followed by exactly one terminal `.ensureResult` (or
    /// `.error`). `which` selects the asset family (Phase 3: only `"speech"`).
    case ensureAssets(which: String)
    /// Forward-compat catch-all: any unknown `type` lands here.
    case unknown
}

// MARK: - Outbound (sidecar → Rust core)

/// Responses sent from the sidecar up to the Rust core.
enum AppleResponse: Equatable {
    /// Result of a `probe`: every boolean reflects a real framework query.
    case probeResult(
        speechAvailable: Bool,
        foundationAvailable: Bool,
        osOk: Bool,
        appleIntelligence: Bool,
        speechAssetsInstalled: Bool
    )
    /// Result of a `summarize`: the generated summary text.
    case summarizeResult(text: String)
    /// Result of a `transcribe`: the finalized transcript text and an OPTIONAL
    /// confidence in `[0, 1]`. `confidence` is `nil` (encoded as JSON null) when
    /// the SDK does not report one — never fabricated (No-Fake-State). Mirrors the
    /// Rust `Option<f32>`.
    case transcribeResult(text: String, confidence: Double?)
    /// Result of an `embedBatch`: one embedding vector per input text, in the
    /// SAME order. Each vector is 512-d (NLEmbedding sentence embedding).
    case embedResult(vectors: [[Float]])
    /// Incremental progress for an `ensureAssets` install: a real fraction in
    /// `[0, 1]`. Emitted zero or more times BEFORE the terminal `.ensureResult`.
    case progress(fraction: Double)
    /// Terminal result of an `ensureAssets` install: whether the assets are now
    /// installed. Emitted exactly once (on success) after any `.progress` lines.
    case ensureResult(installed: Bool)
    /// A failure or degraded condition, carried as a human-readable message.
    case error(message: String)
    /// Forward-compat catch-all: any unknown `type` lands here.
    case unknown
}

// MARK: - Coding keys (exact camelCase, matching the shared fixtures)

/// One flat key-space covering every field across every message.
private enum WireKey: String, CodingKey {
    case type
    // error
    case message
    // summarize (request) / summarizeResult (response) / transcribeResult (response)
    case text
    case instruction
    case maxTokens
    // transcribe (request)
    case pcmBase64
    case locale
    // transcribeResult (response)
    case confidence
    // embedBatch (request)
    case texts
    // embedResult (response)
    case vectors
    // probeResult
    case speechAvailable
    case foundationAvailable
    case osOk
    case appleIntelligence
    case speechAssetsInstalled
    // ensureAssets (request)
    case which
    // progress (response)
    case fraction
    // ensureResult (response)
    case installed
}

// MARK: - Inbound decoding

extension AppleRequest: Decodable {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: WireKey.self)
        // An unknown or absent `type` must degrade to `.unknown`, never throw.
        guard let type = try? c.decode(String.self, forKey: .type) else {
            self = .unknown
            return
        }
        switch type {
        case "probe":
            self = .probe
        case "summarize":
            self = .summarize(
                text: try c.decode(String.self, forKey: .text),
                instruction: try c.decode(String.self, forKey: .instruction),
                maxTokens: try c.decode(Int.self, forKey: .maxTokens)
            )
        case "transcribe":
            self = .transcribe(
                pcmBase64: try c.decode(String.self, forKey: .pcmBase64),
                locale: try c.decode(String.self, forKey: .locale)
            )
        case "embedBatch":
            self = .embedBatch(texts: try c.decode([String].self, forKey: .texts))
        case "shutdown":
            self = .shutdown
        case "ensureAssets":
            self = .ensureAssets(which: try c.decode(String.self, forKey: .which))
        default:
            self = .unknown
        }
    }
}

// MARK: - Inbound encoding
//
// Not used on the hot path (the sidecar only RECEIVES requests), but implemented
// so ProtocolTests can round-trip fixtures symmetrically.

extension AppleRequest: Encodable {
    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: WireKey.self)
        switch self {
        case .probe:
            try c.encode("probe", forKey: .type)
        case let .summarize(text, instruction, maxTokens):
            try c.encode("summarize", forKey: .type)
            try c.encode(text, forKey: .text)
            try c.encode(instruction, forKey: .instruction)
            try c.encode(maxTokens, forKey: .maxTokens)
        case let .transcribe(pcmBase64, locale):
            try c.encode("transcribe", forKey: .type)
            try c.encode(pcmBase64, forKey: .pcmBase64)
            try c.encode(locale, forKey: .locale)
        case let .embedBatch(texts):
            try c.encode("embedBatch", forKey: .type)
            try c.encode(texts, forKey: .texts)
        case .shutdown:
            try c.encode("shutdown", forKey: .type)
        case let .ensureAssets(which):
            try c.encode("ensureAssets", forKey: .type)
            try c.encode(which, forKey: .which)
        case .unknown:
            try c.encode("unknown", forKey: .type)
        }
    }
}

// MARK: - Outbound encoding (the load-bearing shape)

extension AppleResponse: Encodable {
    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: WireKey.self)
        switch self {
        case let .probeResult(speechAvailable, foundationAvailable, osOk, appleIntelligence, speechAssetsInstalled):
            try c.encode("probeResult", forKey: .type)
            try c.encode(speechAvailable, forKey: .speechAvailable)
            try c.encode(foundationAvailable, forKey: .foundationAvailable)
            try c.encode(osOk, forKey: .osOk)
            try c.encode(appleIntelligence, forKey: .appleIntelligence)
            try c.encode(speechAssetsInstalled, forKey: .speechAssetsInstalled)
        case let .summarizeResult(text):
            try c.encode("summarizeResult", forKey: .type)
            try c.encode(text, forKey: .text)
        case let .transcribeResult(text, confidence):
            try c.encode("transcribeResult", forKey: .type)
            try c.encode(text, forKey: .text)
            // Encode confidence as an explicit JSON null when absent (matches the
            // Rust `Option<f32>` shape) — never omit the key, never fabricate.
            if let confidence {
                try c.encode(confidence, forKey: .confidence)
            } else {
                try c.encodeNil(forKey: .confidence)
            }
        case let .embedResult(vectors):
            try c.encode("embedResult", forKey: .type)
            try c.encode(vectors, forKey: .vectors)
        case let .progress(fraction):
            try c.encode("progress", forKey: .type)
            try c.encode(fraction, forKey: .fraction)
        case let .ensureResult(installed):
            try c.encode("ensureResult", forKey: .type)
            try c.encode(installed, forKey: .installed)
        case let .error(message):
            try c.encode("error", forKey: .type)
            try c.encode(message, forKey: .message)
        case .unknown:
            try c.encode("unknown", forKey: .type)
        }
    }
}

// MARK: - Outbound decoding
//
// Implemented so ProtocolTests can decode the outbound fixtures (probe_result,
// error) and assert the expected case. The sidecar itself only SENDS responses,
// so this is test-facing, but it keeps the conformance symmetric.

extension AppleResponse: Decodable {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: WireKey.self)
        guard let type = try? c.decode(String.self, forKey: .type) else {
            self = .unknown
            return
        }
        switch type {
        case "probeResult":
            self = .probeResult(
                speechAvailable: try c.decode(Bool.self, forKey: .speechAvailable),
                foundationAvailable: try c.decode(Bool.self, forKey: .foundationAvailable),
                osOk: try c.decode(Bool.self, forKey: .osOk),
                appleIntelligence: try c.decode(Bool.self, forKey: .appleIntelligence),
                speechAssetsInstalled: try c.decode(Bool.self, forKey: .speechAssetsInstalled)
            )
        case "summarizeResult":
            self = .summarizeResult(text: try c.decode(String.self, forKey: .text))
        case "transcribeResult":
            // `confidence` may be present-and-numeric, JSON null, or absent — all
            // three degrade to `nil` via decodeIfPresent (null → nil).
            self = .transcribeResult(
                text: try c.decode(String.self, forKey: .text),
                confidence: try c.decodeIfPresent(Double.self, forKey: .confidence)
            )
        case "embedResult":
            self = .embedResult(vectors: try c.decode([[Float]].self, forKey: .vectors))
        case "progress":
            self = .progress(fraction: try c.decode(Double.self, forKey: .fraction))
        case "ensureResult":
            self = .ensureResult(installed: try c.decode(Bool.self, forKey: .installed))
        case "error":
            self = .error(message: try c.decode(String.self, forKey: .message))
        default:
            self = .unknown
        }
    }
}

File: frontend/src-tauri/src/apple/protocol.rs
//! Apple helper IPC wire protocol.
//!
//! Transport is NDJSON: one JSON object per line, UTF-8. Field names are exact
//! camelCase (Swift `Codable` on the sidecar decodes/encodes the very same
//! `fixtures/*.json` files), so the wire shape here and the fixtures must stay
//! byte-compatible.
//!
//! Direction:
//! - [`AppleRequest`]  — Rust core → apple-helper sidecar
//! - [`AppleResponse`] — apple-helper sidecar → Rust core
//!
//! Forward-compatibility: both enums carry a `#[serde(other)] Unknown`
//! catch-all, so an unrecognized `type` deserializes to `Unknown` rather than
//! erroring — later phases add new variants without breaking older peers.

use serde::{Deserialize, Serialize};

/// Requests: Rust core → apple-helper sidecar. NDJSON, camelCase.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum AppleRequest {
    /// Ask the sidecar to report Apple STT/LLM availability.
    Probe,
    /// Ask the sidecar to shut down cleanly.
    Shutdown,
    /// Ask the sidecar to summarize `text` under `instruction` (FoundationModels).
    /// The variant-level rename maps `max_tokens` → `maxTokens` on the wire;
    /// `text`/`instruction` are already valid camelCase.
    #[serde(rename_all = "camelCase")]
    Summarize {
        text: String,
        instruction: String,
        max_tokens: u32,
    },
    /// Ask the sidecar to ensure on-device assets (e.g. Speech models) are
    /// installed, streaming zero+ `progress` replies then a terminal
    /// `ensureResult`. Single lowercase field — no rename needed.
    EnsureAssets { which: String },
    /// Ask the sidecar to transcribe one PCM segment via SpeechAnalyzer.
    /// The variant-level rename maps `pcm_base64` → `pcmBase64` on the wire;
    /// `locale` is already valid camelCase. `pcmBase64` is base64 of
    /// little-endian Float32 16 kHz mono PCM.
    #[serde(rename_all = "camelCase")]
    Transcribe { pcm_base64: String, locale: String },
    /// Ask the sidecar to embed a batch of `texts` via on-device NLEmbedding.
    /// Reply is a single [`AppleResponse::EmbedResult`] with one vector per input
    /// (same order) on success, or [`AppleResponse::Error`] on any failure. The
    /// single lowercase field `texts` is already valid camelCase — no rename.
    EmbedBatch { texts: Vec<String> },
    /// Forward-compat catch-all: any unknown `type` lands here.
    #[serde(other)]
    Unknown,
}

/// Responses: apple-helper sidecar → Rust core. Distinct `type` per reply.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum AppleResponse {
    /// Result of a [`AppleRequest::Probe`]: the availability snapshot.
    #[serde(rename_all = "camelCase")]
    ProbeResult {
        speech_available: bool,
        foundation_available: bool,
        os_ok: bool,
        apple_intelligence: bool,
        speech_assets_installed: bool,
    },
    /// A failure the sidecar wants surfaced with a human-readable reason.
    Error { message: String },
    /// Result of a [`AppleRequest::Summarize`]: the generated summary text.
    /// Single field `text` is already valid camelCase, so no rename is needed.
    SummarizeResult { text: String },
    /// A streamed asset-download progress tick (`0.0..=1.0`). Zero or more of
    /// these precede a terminal [`AppleResponse::EnsureResult`].
    Progress { fraction: f64 },
    /// Terminal result of a [`AppleRequest::EnsureAssets`]: whether the assets
    /// are now installed.
    EnsureResult { installed: bool },
    /// Result of a [`AppleRequest::Transcribe`]: the recognized `text` and an
    /// optional `confidence` (`0.0..=1.0`). `text`/`confidence` are already
    /// valid camelCase; `Option<f32>` serializes `null` when `None`, matching
    /// the Swift `Double?`.
    TranscribeResult { text: String, confidence: Option<f32> },
    /// Result of a [`AppleRequest::EmbedBatch`]: one embedding `vectors` entry per
    /// input text, in the SAME order. Each vector is 512-d (NLEmbedding sentence
    /// embedding). Single field `vectors` is already valid camelCase — no rename.
    EmbedResult { vectors: Vec<Vec<f32>> },
    /// Forward-compat catch-all: any unknown `type` lands here.
    #[serde(other)]
    Unknown,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::Value;

    /// Parse both sides to `serde_json::Value` and compare — order-insensitive.
    fn assert_semantic_eq(a: &str, b: &str) {
        let va: Value = serde_json::from_str(a).expect("lhs is valid json");
        let vb: Value = serde_json::from_str(b).expect("rhs is valid json");
        assert_eq!(va, vb, "semantic JSON mismatch\n  lhs: {a}\n  rhs: {b}");
    }

    // ---- Request fixtures (Rust → sidecar) ----

    #[test]
    fn probe_request_roundtrips() {
        let fixture = include_str!("fixtures/probe.json");
        let parsed: AppleRequest =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleRequest");
        assert_eq!(parsed, AppleRequest::Probe);
        assert_ne!(parsed, AppleRequest::Unknown);
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    #[test]
    fn summarize_request_roundtrips() {
        let fixture = include_str!("fixtures/summarize.json");
        let parsed: AppleRequest =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleRequest");
        match &parsed {
            AppleRequest::Summarize {
                text,
                instruction,
                max_tokens,
            } => {
                assert!(text.contains("ship the release on Friday"));
                assert_eq!(
                    instruction,
                    "Summarize the key decisions and action items from this meeting transcript."
                );
                assert_eq!(*max_tokens, 512);
            }
            other => panic!("expected Summarize, got {other:?}"),
        }
        assert_ne!(parsed, AppleRequest::Unknown);
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    #[test]
    fn shutdown_request_roundtrips() {
        let fixture = include_str!("fixtures/shutdown.json");
        let parsed: AppleRequest =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleRequest");
        assert_eq!(parsed, AppleRequest::Shutdown);
        assert_ne!(parsed, AppleRequest::Unknown);
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    #[test]
    fn ensure_assets_request_roundtrips() {
        let fixture = include_str!("fixtures/ensure_assets.json");
        let parsed: AppleRequest =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleRequest");
        match &parsed {
            AppleRequest::EnsureAssets { which } => assert_eq!(which, "speech"),
            other => panic!("expected EnsureAssets, got {other:?}"),
        }
        assert_ne!(parsed, AppleRequest::Unknown);
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    // ---- Response fixtures (sidecar → Rust) ----

    #[test]
    fn probe_result_response_roundtrips() {
        let fixture = include_str!("fixtures/probe_result.json");
        let parsed: AppleResponse =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleResponse");
        assert_ne!(parsed, AppleResponse::Unknown, "fixture must be a known variant");
        match &parsed {
            AppleResponse::ProbeResult {
                speech_available,
                foundation_available,
                os_ok,
                apple_intelligence,
                speech_assets_installed,
            } => {
                assert!(*speech_available);
                assert!(*foundation_available);
                assert!(*os_ok);
                assert!(*apple_intelligence);
                assert!(!*speech_assets_installed);
            }
            other => panic!("expected ProbeResult, got {other:?}"),
        }
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    #[test]
    fn error_response_roundtrips() {
        let fixture = include_str!("fixtures/error.json");
        let parsed: AppleResponse =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleResponse");
        assert_ne!(parsed, AppleResponse::Unknown, "fixture must be a known variant");
        match &parsed {
            AppleResponse::Error { message } => {
                assert_eq!(message, "Apple Intelligence is not enabled");
            }
            other => panic!("expected Error, got {other:?}"),
        }
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    #[test]
    fn summarize_result_response_roundtrips() {
        let fixture = include_str!("fixtures/summarize_result.json");
        let parsed: AppleResponse =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleResponse");
        assert_ne!(parsed, AppleResponse::Unknown, "fixture must be a known variant");
        match &parsed {
            AppleResponse::SummarizeResult { text } => {
                assert!(text.contains("ship the release on Friday"));
            }
            other => panic!("expected SummarizeResult, got {other:?}"),
        }
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    #[test]
    fn progress_response_roundtrips() {
        let fixture = include_str!("fixtures/progress.json");
        let parsed: AppleResponse =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleResponse");
        assert_ne!(parsed, AppleResponse::Unknown, "fixture must be a known variant");
        match &parsed {
            AppleResponse::Progress { fraction } => assert_eq!(*fraction, 0.42),
            other => panic!("expected Progress, got {other:?}"),
        }
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    #[test]
    fn ensure_result_response_roundtrips() {
        let fixture = include_str!("fixtures/ensure_result.json");
        let parsed: AppleResponse =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleResponse");
        assert_ne!(parsed, AppleResponse::Unknown, "fixture must be a known variant");
        match &parsed {
            AppleResponse::EnsureResult { installed } => assert!(*installed),
            other => panic!("expected EnsureResult, got {other:?}"),
        }
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    #[test]
    fn transcribe_request_roundtrips() {
        let fixture = include_str!("fixtures/transcribe.json");
        let parsed: AppleRequest =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleRequest");
        match &parsed {
            AppleRequest::Transcribe { pcm_base64, locale } => {
                assert_eq!(pcm_base64, "AAAAAAAAAAAAAAAAAAAAAA==");
                assert_eq!(locale, "en-US");
            }
            other => panic!("expected Transcribe, got {other:?}"),
        }
        assert_ne!(parsed, AppleRequest::Unknown);
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    #[test]
    fn transcribe_result_response_roundtrips() {
        let fixture = include_str!("fixtures/transcribe_result.json");
        let parsed: AppleResponse =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleResponse");
        assert_ne!(parsed, AppleResponse::Unknown, "fixture must be a known variant");
        match &parsed {
            AppleResponse::TranscribeResult { text, confidence } => {
                assert_eq!(text, "Let's start with the roadmap.");
                assert_eq!(*confidence, Some(0.94));
            }
            other => panic!("expected TranscribeResult, got {other:?}"),
        }
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    #[test]
    fn embed_batch_request_roundtrips() {
        let fixture = include_str!("fixtures/embed_batch.json");
        let parsed: AppleRequest =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleRequest");
        match &parsed {
            AppleRequest::EmbedBatch { texts } => {
                assert_eq!(texts.len(), 2);
                assert!(texts[0].contains("ship the release on Friday"));
                assert!(texts[1].contains("finish the API by Thursday"));
            }
            other => panic!("expected EmbedBatch, got {other:?}"),
        }
        assert_ne!(parsed, AppleRequest::Unknown);
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    #[test]
    fn embed_result_response_roundtrips() {
        let fixture = include_str!("fixtures/embed_result.json");
        let parsed: AppleResponse =
            serde_json::from_str(fixture).expect("fixture deserializes to AppleResponse");
        assert_ne!(parsed, AppleResponse::Unknown, "fixture must be a known variant");
        match &parsed {
            AppleResponse::EmbedResult { vectors } => {
                assert_eq!(vectors.len(), 2);
                assert_eq!(vectors[0], vec![0.1, 0.2, 0.3]);
                assert_eq!(vectors[1], vec![-0.4, 0.5, -0.6]);
            }
            other => panic!("expected EmbedResult, got {other:?}"),
        }
        let reserialized = serde_json::to_string(&parsed).expect("serializes back");
        assert_semantic_eq(fixture, &reserialized);
    }

    // ---- camelCase mapping assertions ----

    #[test]
    fn probe_result_serializes_camel_case_tag_and_fields() {
        let value = serde_json::to_value(AppleResponse::ProbeResult {
            speech_available: true,
            foundation_available: false,
            os_ok: true,
            apple_intelligence: false,
            speech_assets_installed: true,
        })
        .unwrap();
        assert_eq!(value["type"], "probeResult");
        assert_eq!(value["speechAvailable"], true);
        assert_eq!(value["foundationAvailable"], false);
        assert_eq!(value["osOk"], true);
        assert_eq!(value["appleIntelligence"], false);
        assert_eq!(value["speechAssetsInstalled"], true);
    }

    // ---- Forward-compatibility: unknown `type` → Unknown ----

    #[test]
    fn unknown_request_type_deserializes_to_unknown() {
        // A type the protocol will never define — must degrade to `Unknown`,
        // never error. (Deliberately not a real/future variant name; `transcribe`
        // is now a real request variant, so it can't stand in for "unknown".)
        let line = r#"{"type":"definitelyNotARealAppleRequestType","x":1}"#;
        let parsed: AppleRequest = serde_json::from_str(line).expect("must not error");
        assert_eq!(parsed, AppleRequest::Unknown);
    }

    #[test]
    fn unknown_response_type_deserializes_to_unknown() {
        // A type the protocol will never define — must degrade to `Unknown`,
        // never error. (Deliberately not a real/future variant name.)
        let line = r#"{"type":"definitelyNotARealAppleResponseType","text":"hi"}"#;
        let parsed: AppleResponse = serde_json::from_str(line).expect("must not error");
        assert_eq!(parsed, AppleResponse::Unknown);
    }
}
